### PR TITLE
[FEATURE] #9 메인-홈 화면의 편의점별 행사 상품 UI 추가

### LINKED ISSUE #9: [FEATURE] 메인-홈 화면과 관련된 기획 및 UI 완성
### 📌 내용
메인화면의 홈 탭과 관련된 기획서 작성 및 UI 구현 완료 목표

### 🛠️ To Do
- [x]  화면 기획 및 문서 작성
- [x]  필요한 클래스 및 API 뽑아두기
- [x]  데이터 Modeling
- [x]  필요한 string → resource에 저장
- [x]  상품 ViewHolder
- [x]  리뷰 ViewHolder
- [x]  상품 리스트 ViewHolder
- [x]  상품 리스트
- [x]  리뷰 리스트
- [x] 편의점별 이벤트 상품 리스트
- [ ] 리스트 스크롤 시에도 각 item 의 state가 유지되도록 처리

### 🪴 관련 Ticket
- [[AOS] 메인-홈 화면 기획 및 UI 완성](https://www.notion.so/yapp-workspace/AOS-UI-aad9c427eef14a989cdce3ebc56cd967?pvs=4)

### PR DESCRIPTION
**Pull request by aurora32s**
#### 📌 내용
메인(홈) 화면에서 `편의점별 행사 상품` 리스트와 관련된 UiModel / ViewHolder / Layout 코드 추가

#### 🛠 Done
- [x] 편의점별 행사 상품과 관련된 UiModel 추가
- [x] 편의점별 행사 상품 Layout 추가
- [x] 편의점별 행사 상품 ViewHolder 추가
- [x] 편의점별로 ViewPager 적용
- [x] 화면에 보이는 편의점 리스트에 따라 Tab이 변경되도록 처리

#### 🏠 중심 코드
- EventByStoresViewHolder

#### 🪴 참고 UI

https://github.com/YAPP-Github/22nd-Android-Team-1-Android/assets/22411296/fbbdb749-4534-4cce-83a9-792c813c7ca2

#### 🙏 변경사항
```
peonlee_medium_selector
```
- text 가운데 정렬
- text의 left margin 재거
- imageView의 right margin 제거
- horizontal padding이 필요한 경우에는 사용하는 쪽에서 padding 적용하는 방법을 사용할 수 있음

#### 🔖 관련 Issue
Closed #9

### CODE REVIEW

File: feature/home/src/main/java/com/peonlee/home/adapter/viewholder/product/EventByStoresViewHolder.kt [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import com.peonlee.home.model.product.EventByStoresUiModel
+import com.peonlee.model.type.StoreType
+import com.peonlee.core.ui.R as UiResource
+import com.peonlee.home.R as HomeResource
+
+class EventByStoresViewHolder(
+    private val binding: ListItemEventStoresBinding
+) : CommonViewHolder<EventByStoresUiModel>(binding) {
+    private var productsByStoreAdapter = ProductsByStoreAdapter()
+    private var tabLayoutMediator: TabLayoutMediator? = null
```
Comment (jiwon2724) [inline feature/home/src/main/java/com/peonlee/home/adapter/viewholder/product/EventByStoresViewHolder.kt]: 궁금한 부분입니다!
`tabLayoutMediator`가 null일 경우가 있을까요~?.? 

만약 없다면, `tabLayoutMediator`을 non-null로 만들고 
`tabLayoutMediator`참조하는 부분에서 safe call을 사용 안 해도 괜찮을 것 같습니다!

File: feature/home/src/main/java/com/peonlee/home/adapter/viewholder/product/EventByStoresViewHolder.kt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            binding.layoutStoreTab.removeOnTabSelectedListener(onTabSelectedListener)
+            tabLayoutMediator?.detach()
+        }
+    }
+
+    override fun onBindView(item: EventByStoresUiModel) {
+        productsByStoreAdapter.submitList(item.stores)
+    }
+
+    companion object {
```
Comment (jiwon2724) [inline feature/home/src/main/java/com/peonlee/home/adapter/viewholder/product/EventByStoresViewHolder.kt:69]: 단순 궁금증입니다!

`companion object`를 사용 안했을 경우에 `getBackgroundByStore`, `getIconByStore`를
`init`이나 `onTabSelected`에서 참조를 못할까요?.?

**Review by jiwon2724 — COMMENTED** [review]
고생많으셨습니다~~!!!:raised_hands::raised_hands:
궁금한 부분 코멘트 남겼습니다! 확인 부탁드릴게요~~

File: feature/home/src/main/java/com/peonlee/home/adapter/viewholder/product/EventByStoresViewHolder.kt (hunk 2 above)
Comment (aurora32s) [reply feature/home/src/main/java/com/peonlee/home/adapter/viewholder/product/EventByStoresViewHolder.kt:69]: 참조할 수 있을거예요.
`getBackgroundByStore`과 `getIconByStore`가 클래스 내부의 프로퍼티를 사용하지 않기 때문에 각 객체별로 메서드를 가지고 있는 것보다 static으로 선언해두는 쪽이 맞을듯 하여 companion object 내에 두게 되었습니다. :-)

File: feature/home/src/main/java/com/peonlee/home/adapter/viewholder/product/EventByStoresViewHolder.kt [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                )
+                tabItem.root.text = store.storeName
+                tabItem.root.setIcon(getIconByStore(store))
+                tab.customView = tabItem.root
+            }
+        }
+        itemView.doOnAttach {
+            binding.layoutStoreTab.addOnTabSelectedListener(onTabSelectedListener)
+            tabLayoutMediator?.attach()
+        }
```
Comment (myung6024) [inline feature/home/src/main/java/com/peonlee/home/adapter/viewholder/product/EventByStoresViewHolder.kt]:
```
public inline fun View.doOnAttach(crossinline action: (view: View) -> Unit) {
    if (ViewCompat.isAttachedToWindow(this)) {
        action(this)
    } else {
        addOnAttachStateChangeListener(object : View.OnAttachStateChangeListener {
            override fun onViewAttachedToWindow(view: View) {
                removeOnAttachStateChangeListener(this)
                action(view)
            }

            override fun onViewDetachedFromWindow(view: View) {}
        })
    }
}
```

doOnAttach는 한번만 호출되는 extension이라서 아마 의도한 동작과 다르게 동작할 것 같아요

File: feature/home/src/main/java/com/peonlee/home/adapter/viewholder/product/EventByStoresViewHolder.kt (hunk 1 above)
Comment (aurora32s) [reply feature/home/src/main/java/com/peonlee/home/adapter/viewholder/product/EventByStoresViewHolder.kt]: 오호. 좋은 의견 감사드립니다. 😌 해당 부분 수정해서 반영하도록 하겠습니다.

File: feature/home/src/main/java/com/peonlee/home/adapter/viewholder/product/ProductsByStoreViewHolder.kt [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            }
+            layoutProducts.addView(
+                ProductViewHolder(
+                    ListItemProductBinding.inflate(LayoutInflater.from(itemView.context))
+                ).also {
+                    it.onBindView(product)
+                }.itemView,
+                params
+            )
+        }
```
Comment (myung6024) [inline feature/home/src/main/java/com/peonlee/home/adapter/viewholder/product/ProductsByStoreViewHolder.kt]: 이렇게 되면 onBindView될떄마다 addView가 계속 추가 되지 않을까요?

File: feature/home/src/main/java/com/peonlee/home/HomeFragment.kt [hunk 5]
```diff
@@ -19,6 +23,20 @@ class HomeFragment : BaseFragment<FragmentHomeBinding>() {
...
+                    val viewHolder = binding.layoutHome.getChildViewHolder(view)
+                    (viewHolder as? EventByStoresViewHolder)?.doOnAttach()
+                }
+
+                override fun onChildViewDetachedFromWindow(view: View) {
+                    val viewHolder = binding.layoutHome.getChildViewHolder(view)
+                    (viewHolder as? EventByStoresViewHolder)?.doOnDetach()
+                }
+            }
+        )
```
Comment (myung6024) [inline feature/home/src/main/java/com/peonlee/home/HomeFragment.kt]: select listener 등록을 attach detach로 관리하기 위한 로직이 코드 복잡도를 높이는 것 같아요. 큰 이점이 없다면 가독성을 우선하로 하는 것도 좋을 것 같아요

File: feature/home/src/main/java/com/peonlee/home/HomeFragment.kt [hunk 6]
```diff
@@ -19,6 +23,20 @@ class HomeFragment : BaseFragment<FragmentHomeBinding>() {
         val adapter = HomeAdapter()
         binding.layoutHome.adapter = adapter
```
Comment (myung6024) [inline feature/home/src/main/java/com/peonlee/home/HomeFragment.kt]: layoutHome은 recycelrview일까요? rvHome이 되어야 할 것 같아서요

File: feature/home/src/main/res/layout/list_item_conditional_products.xml [hunk 7]
```diff
@@ -20,9 +20,9 @@
         android:id="@+id/btn_more_products"
         android:layout_width="0dp"
         android:layout_height="wrap_content"
-        android:layout_marginHorizontal="@dimen/item_conditional_products_button_horizontal_margin"
-        android:layout_marginTop="@dimen/item_conditional_products_button_top_margin"
-        android:layout_marginBottom="@dimen/item_conditional_products_button_bottom_margin"
+        android:layout_marginHorizontal="@dimen/item_button_horizontal_margin"
+        android:layout_marginTop="@dimen/item_button_top_margin"
+        android:layout_marginBottom="@dimen/item_button_bottom_margin"
```
Comment (myung6024) [inline feature/home/src/main/res/layout/list_item_conditional_products.xml:25]: 모든 margin에 대해서 dimen을 정의가 되어야할까욤? 어떤 이점이 있을까요?

File: feature/home/src/main/java/com/peonlee/home/HomeFragment.kt (hunk 6 above)
Comment (aurora32s) [reply feature/home/src/main/java/com/peonlee/home/HomeFragment.kt]: 넵. 수정 후 반영하도록 하겠습니다. 👍🏻

File: feature/home/src/main/java/com/peonlee/home/HomeFragment.kt (hunk 5 above)
Comment (aurora32s) [reply feature/home/src/main/java/com/peonlee/home/HomeFragment.kt]: 넵. 확인하였습니다.
해당 부분은 이전에 말씀드렸듯이, ViewHolder 내에서 한번만 적용하도록 수정 후 반영하도록 하겠습니다. 👍🏻

File: feature/home/src/main/java/com/peonlee/home/adapter/viewholder/product/ProductsByStoreViewHolder.kt (hunk 4 above)
Comment (aurora32s) [reply feature/home/src/main/java/com/peonlee/home/adapter/viewholder/product/ProductsByStoreViewHolder.kt]: 해당 부분은 만약 layoutProducts 하위에 view가 있다면 전부 제거하고 다시 추가하는 방법으로 구현해보려는데 괜찮을까요?

File: feature/home/src/main/res/layout/list_item_conditional_products.xml (hunk 7 above)
Comment (aurora32s) [reply feature/home/src/main/res/layout/list_item_conditional_products.xml:25]: 먼저 xml에 dp를 바로 적으면 warning이 떠서 전부 dimen 파일로 옮겨둔 점이 있어요.
그리고 다른 layout 들과 비교해서도 딱 정해진 padding 규격이 있는게 아니여서
```
예를 들어, `2 / 4 / 8 / 16 dp의 padding들만 사용된다.` 와 같은 규격
```
하나의 layout 별로 dimen value들을 모두 만들어두었습니다.ㅎㅎ

이 부분은 추후에 화면들이 다 완성되고 난 이후에 따로 Refactoring 기간을 가져
정리를 해보아도 괜찮을까요? 😢

File: feature/home/src/main/java/com/peonlee/home/adapter/viewholder/product/ProductsByStoreViewHolder.kt (hunk 4 above)
Comment (myung6024) [reply feature/home/src/main/java/com/peonlee/home/adapter/viewholder/product/ProductsByStoreViewHolder.kt]: 넵넵

**Review by jiwon2724 — APPROVED** [review]
고생 많으셨습니다~~~!! 역최개! :thumbsup::thumbsup:

**Review by myung6024 — APPROVED** [review]
수고하셨습니다~!
